pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/SegmentDeletionManager.java

```diff
@@ -74,12 +75,17 @@ public class SegmentDeletionManager {
     RETENTION_DATE_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
   }
 
+  public interface SegmentDeletionHook {
+    void process(String tableName, List<String> segmentsToDelete);
+  }
+
   private final ScheduledExecutorService _executorService;
   private final String _dataDir;
   private final String _helixClusterName;
   private final HelixAdmin _helixAdmin;
   private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
   private final long _defaultDeletedSegmentsRetentionMs;
+  private final ConcurrentLinkedQueue<SegmentDeletionHook> _segmentDeletionHooks;
```

Comment:
We probably don't need a concurrent queue since the register should happen before the start (we might want to introduce `start()` to start the background task)

Reply:
For listener model, we usually register the listeners before starting the service. I don't think we plan to make the listener registering dynamic after controller is started. Another way is to simply make `register()` `synchronized` since the access is also in synchronized block to avoid the overhead of concurrent queue.
This is optional as the overhead of using a concurrent queue for very few listeners should also be very small.